minio/select_object.py

```diff
@@ -0,0 +1,219 @@
+        header_byte_parsed += 1
+        value_string_byte_length \
+        = byte_int(header[header_byte_parsed: header_byte_parsed+2])
+        header_byte_parsed += 2
+        header_value \
+        = header[header_byte_parsed:header_byte_parsed+value_string_byte_length]
+        header_byte_parsed += value_string_byte_length
+        header_map[header_name.decode("utf-8").lstrip(":")] = \
+            header_value.decode("utf-8").lstrip(":")
+    return header_map
+
+
+
+def parse_message(header_map, payload, payload_length, record, stat):
+    '''
+    Parses the message
+    '''
+    if header_map["message-type"] == 'error':
+        error = header_map["error-code"] + ":\"" +\
+                header_map["error-message"] + "\""
+        record.write(error)
+    if header_map["message-type"] == 'event':
+        # Fetch the content-type
+        content_type = header_map["content-type"]
+        # Fetch the event-type
+        event_type = header_map["event-type"]
+        if event_type == 'Records':
+            rec = payload[0:payload_length]
+            record.write(rec)
+        elif  event_type == 'Progress':
+            if content_type == "text/xml":
+                progress = payload[0:payload_length]
+                stat.write(progress)
+        elif event_type == "Stats":
+            if content_type == "text/xml":
+                stats = payload[0:payload_length]
+                stat.write(stats)
+
+def extract_component(response, read_buffer, bytes_parsed,
+                      size, chunked_message):
+    '''
+    Fetches all the individual component of message namely
+    1. total_byte_length
+    2. headers_byte_length
+    3. prelude_crc
+    4. header
+    5. payload
+    6. message_crc
+    '''
+    flag = False
+    if read_buffer - bytes_parsed < size:
+        value = chunked_message[bytes_parsed:bytes_parsed + \
+                                (read_buffer - bytes_parsed)]
+        rem_bytes = response.read(size- (read_buffer - bytes_parsed))
+        if len(rem_bytes) > 0:
+            chunked_message = response.read(read_buffer)
+            flag = True
+        value += rem_bytes
+    else:
+        value = chunked_message[bytes_parsed:bytes_parsed + size]
+    if flag:
+        bytes_parsed = 0
+    else:
+        bytes_parsed += size
+    return  value, bytes_parsed, chunked_message
+
+
+def extract_message(response):
+    '''
+    Process the response sent from server.
+    Refer https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html
+    '''
+    read_buffer = READ_BUFFER_SELECT
+    record = io.BytesIO()
+    stats = io.BytesIO()
+    # Response read in chunks of read_buffer
+    chunked_message = response.read(read_buffer)
+    bytes_parsed = 0
+    while bytes_parsed < read_buffer:
+        #Process total byte length
+        total_byte_length, bytes_parsed, chunked_message = \
+            extract_component(response, read_buffer,
```

Comment:
as `read_buffer` is a constant, we dont need to pass it as a parameter to the calls.